Hardware Unique Key (HUK) generation using the CAAM Master Key Verification Blob (MKVB) for i.MX platforms.

Emantor opened the pull request:
This is a port of PR #3160 to the NXP CAAM driver. The i.MX platform configuration is also modified to enable the MKVB generation for all possible platforms.
This PR:
1. introduces a early_driver initcall which runs before the service initcalls
2. moves the CAAM initialization there
3. implements a MKVB module which retrieves the MKVB
4. implements the platform implementation to retrieve a HUK using the MKVB
5. moves the jobring handoff into the driver_late initcall

This way there is always a jobring available during OP-TEE initialization which is used to retrieve the MKVB. Instead of disabling the complete CAAM driver, the default configuration now disables the crypto driver and runtime jobring access to prevent the SoC hang if clocks are disabled.

cc @clementfaure

jforissier commented on core/drivers/crypto/caam/mkvb/caam_mkvb.c:
```diff
@@ -0,0 +1,69 @@
+#include <stdint.h>
+#include <tee/cache.h>
+#include <string.h>
+#include <mm/core_memprot.h>
```
Minor: alphabetical order preferred

Emantor replied:
Added a fixup commit.

jforissier commented on core/include/drivers/imx_caam_mkvb.h:
```diff
@@ -0,0 +1,21 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright 2020 Pengutronix, Rouven Czerwinski <entwicklung@pengutronix.de>
```
Minor: not the same email address as Author/S-o-b but you can leave it as is if you prefer.

Emantor replied:
We prefer this for file headers, entwicklung (german for development) will reach all developers for questions even if the person may have left the company in the meantime.

jforissier replied:
Das ist für mich kein Problem :wink:

jforissier commented on core/arch/arm/plat-imx/conf.mk:
```diff
@@ -387,16 +387,24 @@ CFG_MMAP_REGIONS ?= 24
 # it with generic crypto API can be enabled.
-CFG_CRYPTO_DRIVER ?= y
+# Currently disabled by default since it requires patching the linux CAAM driver
+# to leave the clocks to the CAAM enabled or removal of the DECO instanziation
```
instantiation

Emantor replied:
Added a fixup commit.

Emantor commented:
> Hi @Emantor ,
> 
> LGTM with a few minor things below. Of course I'm not considering the platform-specific aspects, only general things.
> 
>     * In commit "caam: add function to increment PRIBLOB":
>       > Add a function which increments the PRIBLOB field to 11.
>       
>       
>       Do you mean binary 11? The code has `io_setbits32(ctrl_base + SCFGR, 0x3);`.

Yes, I'll add a 0b prefix while squashing.
> 
>     * Should commit "caam: replace CFG_NXP_CAAM_RUNTIME_JR ifdefs" be moved before "imx: reenable CFG_NXP_CAAM, disable runtime access" perhaps? (it should if it's the latter that causes the build warning/error)

Thats indeed the case. I'll squash the commit into the configuration change.

cneveux commented on core/drivers/crypto/caam/sub.mk:
```diff
@@ -9,4 +9,4 @@ srcs-y += caam_jr.c
 srcs-y += caam_desc.c
 subdirs-$(CFG_NXP_CAAM_HASH_DRV) += hash
-
+subdirs-$(CFG_NXP_CAAM_MKVB_DRV) += mkvb
```
master key derivation is a particular CAAM BLOB operation. It would be better for future blob operation to rename the mkvb to blob (folder like CFG define)

Emantor replied:
Will do.

clementfaure replied:
@Emantor Do you agree on renaming it to ```blob``` instead of ```mkvb``` ?

cneveux commented on core/arch/arm/plat-imx/drivers/imx_huk.c:
```diff
@@ -0,0 +1,24 @@
+static uint8_t stored_key[MKVB_SIZE];
+static bool mkvb_retrieved;
+
+TEE_Result tee_otp_get_hw_unique_key(struct tee_hw_unique_key *hwkey)
```
This function can be directly part of the caam_mkvb.c file as caam_mkvb.c is built only if CFG_NXP_CAAM_MKVB_DRV enabled it.

Emantor replied:
This will work until plat-ls wants to enable the MKVB/BLOB driver, they also implement a `tee_otp_get_hw_unique_key()` function using a TF-A platform service.

cneveux replied:
Valid point but if a platform including CAAM doesn't need/wait to use this specific CAAM operation, the CFG_CFG_NXP_CAAM_MKVB_DRV doesn't have to be enabled in the configuration file.

cneveux commented on core/drivers/crypto/caam/caam_ctrl.c:
```diff
@@ -81,6 +73,20 @@ static TEE_Result crypto_driver_init(void)
 
 early_driver_init(crypto_driver_init);
 
+static TEE_Result crypto_driver_deinit(void)
```
This is already managed by the original driver. The JR ring used during the initialization is released if the CFG_CRYPTO_DRIVER is not enabled, This function is not needed.

Emantor replied:
The deinitialization is moved since I want the following sequence:
1. early_driver_init of the CAAM
2. service initcalls eventually call into the CAAM to request the MKVB
3. other service calls might use the CAAM to initialize RNG
3. driver_init_call late hands over all jobrings to normal world if `CFG_NXP_CAAM_RUNTIME_JR=n`

cneveux replied:
RNG instantiation can not be done outside Secure environment. (it's may be working on some device not all).
You have to think that what is valid for i.MX6Q is not valid for i.MX7 or i.MX8. This is why this driver has been implemented in generic way with specific HAL.

Emantor replied:
We are not leaving the secure environment, I don't understand how this moves the RNG initialization outside of the secure environment. Point 3. was meant that OP-TEE could use the CAAM to initialize its PRNG in case `CFG_NXP_CAAM_RUNTIME_JR=n`, I was not refering to the RNG instantiation of the RNG state handles inside the CAAM.

cneveux commented on core/drivers/crypto/caam/caam_ctrl.c:
```diff
@@ -81,6 +73,20 @@ static TEE_Result crypto_driver_init(void)
 
 early_driver_init(crypto_driver_init);
```
Is it something really necessary? crypto_driver_init can directly called in the tee_cryp_init function (see below) that enabled the crypto before any crypto operations (like HUK generation). The concept is to be able to replace the CAAM driver with another HW crypto driver easly. 

__weak TEE_Result crypto_driver_init(void)
{
	return TEE_SUCCESS;
}


static TEE_Result tee_cryp_init(void)
{
	TEE_Result res = crypto_init();

	/*
	 * If there is a Cryptographic Driver, we need to
	 * call the initialization function here before using
	 * cryptographic operation during the boot
	 * (e.g. HUK generation)
	 */
	if (res == TEE_SUCCESS)
		res = crypto_driver_init();

	if (res) {
		EMSG("Failed to initialize crypto API: %#" PRIx32, res);
		panic();
	}

	plat_rng_init();

	return TEE_SUCCESS;
}

Emantor replied:
I prefer the clarity of using an explicit new initcall which runs before services are initialized. This also allows a clean handover of jobrings when `CFG_NXP_CAAM_RUNTIME_JR=n`.

cneveux replied:
I would let core maintainers to comment and adjudicated on this point.
Using initcall never guaranty the execution order in case of concurrency.

jforissier replied:
@cneveux what do you mean "concurrency"? initcalls do provide some execution order guarantees between different levels, which should be enough in this case?

As to initcalls vs. direct calls, no strong opinion but I tend to prefer the first because they are bit simpler: no weak symbol, no need to include headers and insert explicit calls, each layer remains independent on the others.

cneveux replied:
@jforissier 
My concern is about the initcalls order which raised each time we need to init A before B and finally C because without A, B can not be initialized and C can not work without B. 
In this case, C can be init thru driver_init and to ensure that B is initialized before C, B is init thru early_driver_init. How to ensure that A is initialized first, we will add a early_early_driver_init?

jforissier replied:
It's a matter of having the "right" number of levels. Dismissing the mechanism just because it doesn't scale perfectly well to large numbers is a bit excessive I think. All the more since it's just a matter of naming (initcall levels are just integers in the end).

Currently we have "services" which are initialized first and then "drivers". The current use case shows that a service (crypto) may depend on specific hardware and need a driver. So introducing "early drivers" makes sense. We don't have any use case where such an "early driver" would depend on another driver yet. So your argument is valid in theory, but is it in practice?

cneveux replied:
I will not debate on this point. I'm not rejecting the current proposal, for sure today solution is acceptable. I'm just saying that we may face another problem something later.

cneveux commented on core/arch/arm/plat-imx/conf.mk:
```diff
@@ -387,16 +387,24 @@ CFG_MMAP_REGIONS ?= 24
 # Almost all platforms include CAAM HW Modules, except the
 # ones forced to be disabled
-CFG_NXP_CAAM ?= n
+CFG_NXP_CAAM ?= y
```
I would prefer to not change this file
CFG_NXP_CAAM_RUNTIME_JR is already managed in the crypto_conf.mk, so no need to change it.
CFG_CRYPTO_DRIVER must be enabled only if the CAAM driver is to be supported at runtime to replace the Crypto Software library operations supported in the CAAM Driver.
CFG_NXP_CAAM_MKVB_DRV  should be defined in the crypto_conf.mk file where all CAAM driver features are managed.

Emantor replied:
This PR intentionally disables `CFG_NXP_CAAM_RUNTIME_JR`, to ensure that OP-TEE does not try to access the CAAM during runtime, because this will currently lead to SoC Lockups (see https://github.com/OP-TEE/optee_os/issues/3405 and the discussion in https://github.com/OP-TEE/optee_os/pull/3349).

cneveux replied:
I still don't understand. CFG_NXP_CAAM_RUNTIME_JR is not defined if the CFG_CRYPTO_DRIVER is not equal to 'y'.
If CFG_CRYPTO_DRIVER is not enabled, CAAM driver is not installed to be used at runtime. runtime CAAM driver has not interest if the CFG_CRYPTO_DRIVER is not used.

Emantor replied:
> I still don't understand. CFG_NXP_CAAM_RUNTIME_JR is not defined if the CFG_CRYPTO_DRIVER is not equal to 'y'.
> If CFG_CRYPTO_DRIVER is not enabled, CAAM driver is not installed to be used at runtime. runtime CAAM driver has not interest if the CFG_CRYPTO_DRIVER is not used.

Thats right, I'll remove the explicit `CFG_NXP_CAAM_RUNTIME_JR`.

cneveux replied:
We are doing our best to satisfy (and enable) all i.MX users. It doesn't mean that we have done as reference can correspond to everybody wanted this or this options or want to disable certain parts. Hence, if you want to re-use plat-imx but with specific build options, change your build command line to enable your option.

Emantor replied:
As I said previously  in https://github.com/OP-TEE/optee_os/pull/3349#issue-333011100, upstream OP-TEE is expected to work with an upstream kernel, which is the reason why `CFG_NXP_CAAM` was disabled in https://github.com/OP-TEE/optee_os/pull/3349. However we now have the case where we require the CAAM driver to be used for the HUK generation (and for the initial RNG Seed), but will still run into the clock conflict issue since the kernel will disable the clocks. Therefore I enabled the CAAM driver and disabled `CFG_CRYPTO_DRIVER` which disables the CAAM runtime access. This way upstream OP-TEE can use the CAAM during initialization for the HUK and RNG seed, but won't conflict with the kernel during runtime.

> We are working on the integration of the CAAM driver on referenced platform, but before we have to ensure that enabled platforms are fully validated.
Your change can break the boot of non validated platforms.

Why was `CFG_NXP_CAAM` than enabled by default before? Either way I validated this change at least on i.MX6Q and i.MX6UL, I'd be okay to only enable these platforms.

Regarding the disabled `CFG_CRYPTO_DRIVER`, @clementfaure mentioned in https://github.com/OP-TEE/optee_os/pull/3349#issuecomment-547790617 that there was a patch set for the problem. I am fully in favor of re-enabling `CFG_CRYPTO_DRIVER` for `plat-imx` once this patchset is merged in the upstream kernel.

cneveux replied:
So you are going the same way as us. Conclusion don't touch this configuration. We will do it.
Waiting that, enable it on your side locally either thru the build command line, either by editing this file but not pushing it to upstream. Thanks.

Emantor replied:
> Conclusion don't touch this configuration.

I don't reach the same conclusions. I'd like HUK generation to be available with the next OP-TEE release which is due some time in april. Current kernel looks to be on schedule for a end of march release, I doubt that the CAAM kernel patches can still target v5.6. Consequently the changes could be in 5.7. which is due after the OP-TEE release. The next OP-TEE release can consequently enable `CFG_CRYPTO_DRIVER`.

clementfaure replied:
Hi @Emantor,
We still believe that ```CFG_NXP_CAAM``` should stay disabled for reasons already explained by @cneveux. The CAAM has been disabled for good reasons and we prefer to wait for a proper patch in the kernel before re-enabling it.
In the mean time, I believe this configuration should be done locally to enabled the HUK generation, but not upstreamed.

cneveux commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -602,9 +603,9 @@ enum caam_status caam_jr_init(struct caam_jrcfg *jrcfg)
-#ifdef CFG_NXP_CAAM_RUNTIME_JR
-	itr_add(&jr_privdata->it_handler);
-#endif
+	if (IS_ENABLED(CFG_NXP_CAAM_RUNTIME_JR))
```
Those changes are not needed in the case of adding the HUK generation with the CAAM Blob operation

Emantor replied:
This change is needed to resolve a compile error when `CFG_NXP_CAAM_RUNTIME_JR=n`.

Emantor commented on core/drivers/crypto/caam/mkvb/caam_mkvb.c:
```diff
@@ -0,0 +1,69 @@
+out_buf:
+	caam_free_buf(&buf);
+	/* Increment PRIBLOB to 11 to lock out same MKVB from normal world */
+	caam_hal_ctrl_inc_priblob();
```
Correct, I'd like to make this configurable than. We can argue about the default, I fear that users forget to set the PRIBLOB bits in the bootloader.

Emantor replied:
> Other point, the PRIBLOB is only used in Trusted mode execution, meaning that when same descriptor is executed in normal world, the master key derived is not the same regardless the PRIBLOB settings.

I haven't found more information on this in the SRM. Is this also true for the CAAM contained in older i.MX6 SoCs like i.MX6Q? The SRM does not mention a jobring owner there and at least with HAB the TZ can not be set.
My understanding of the SRM is that the CAAMs Trusted State has more to do with HAB than with the TrustZone accesses, if you have more information, please share.

cneveux replied:
Have a look chapter Blob types differentiated by security state in the SRM.

Emantor replied:
The way I read it is that this non-secure state describes a not HAB booted device and not a processor or memory access from the TrustZone non-secure world. This IMO confirmed by the state being derived from the SNVS security state, which is documented to indicate the HAB status of the SoC and not the TrustZone non-secure state. Furthermore the SRM describes a transition from trusted state to the non-secure state only through the fail state, with no way to get back to trusted state after that.

cneveux replied:
More explanation about the PRIBLOB. On i.MX6Q (and some other legacy devices), the PRIBLOB sets to 0x3 is a workaround because the Trustzone bit is not connected to the CAAM secure state machine. The problem of setting these value is disabling the CAAM BLOB interest in OPTEE because the same level of protection is used regardless the world running.
Hence, setting the PRIBLOB to 0x3 must remain a **platform specific workaround** done before switching to normal OS boot. It can not be part of CAAM driver at all.

I advice you to add a specific function into plat-imx files that is enabled or not with a CFG_XXX device to let other user to enable it if wanted. Then call this function when the OPTEE boot ends.
Does it looks reasonable for you?

Emantor commented on core/drivers/crypto/caam/mkvb/caam_mkvb.c:
```diff
@@ -0,0 +1,69 @@
+
+	res = caam_calloc_align_buf(&buf, 32);
+	if (res != CAAM_NO_ERROR)
+		goto out_buf;
```
Will fix.

Emantor commented on core/drivers/crypto/caam/mkvb/caam_mkvb.c:
```diff
@@ -0,0 +1,69 @@
+
+	memcpy(dest, buf.data, MKVB_SIZE);
+	ret = TEE_SUCCESS;
+out_desc:
```
Will fix.

Emantor commented on core/drivers/crypto/caam/mkvb/caam_mkvb.c:
```diff
@@ -0,0 +1,69 @@
+	cache_operation(TEE_CACHEINVALIDATE, buf.data, MKVB_SIZE);
+	DHEXDUMP(buf.data, MKVB_SIZE);
+
+	memcpy(dest, buf.data, MKVB_SIZE);
```
Ok.

Emantor commented on core/drivers/crypto/caam/mkvb/caam_mkvb.c:
```diff
@@ -0,0 +1,69 @@
+		goto out_desc;
+	}
+
+	cache_operation(TEE_CACHEINVALIDATE, buf.data, MKVB_SIZE);
```
Ok

Emantor commented on core/drivers/crypto/caam/mkvb/caam_mkvb.c:
```diff
@@ -0,0 +1,69 @@
+	jobctx.desc = desc;
+	ret = caam_jr_enqueue(&jobctx, NULL);
+	DMSG("caam ret: %d", ret);
+	if (ret == CAAM_JOB_STATUS) {
```
Ok

Emantor commented on core/drivers/crypto/caam/mkvb/caam_mkvb.c:
```diff
@@ -0,0 +1,69 @@
+	ret = caam_jr_enqueue(&jobctx, NULL);
+	DMSG("caam ret: %d", ret);
+	if (ret == CAAM_JOB_STATUS) {
+		EMSG("HuK generation failed: Job status 0x%08" PRIx32,
```
Ok

Emantor commented on core/drivers/crypto/caam/mkvb/caam_mkvb.c:
```diff
@@ -0,0 +1,69 @@
+	caam_desc_add_word(desc, DESC_HEADER(0));
+	caam_desc_add_word(desc, MKVB_DESC_SEQ_OUT);
+	caam_desc_add_ptr(desc, buf.paddr);
+	cache_operation(TEE_CACHEFLUSH, buf.data, MKVB_SIZE);
```
Ok

Emantor commented on core/drivers/crypto/caam/mkvb/caam_mkvb.c:
```diff
@@ -0,0 +1,69 @@
+	if (res != CAAM_NO_ERROR)
+		goto out_buf;
+
+	desc = caam_calloc_desc(5);
```
Ok

Emantor commented:
I reworked this, but chose to keep the MKVB name instead of switching to BLOB, this way `plat-ls` can choose to not enable the MKVB part of the CAAM functionality. Since a lot of the commits changed, I chose to apply the tags in tandem. I tried to address all review comments, please comment if I missed some.

Emantor commented:
Rebased onto latest master to fix the merge conflicts. @cneveux  please review again.

cneveux commented on core/arch/arm/plat-imx/drivers/imx_priblob.c:
```diff
@@ -0,0 +1,40 @@
+
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, CAAM_BASE, CORE_MMU_PGDIR_SIZE);
+
+static TEE_Result lock_priblob(void)
```
Why implementing this file.
1. You are redefining SCFGR which is already defined in the caam/hal/imx_[6_7/8m]/registers/ctrl_regs.h
2. This function must be part of the caam/hal/common/hal_ctrl.c file
3. We have to ensure that this function is callled before leaving the secure environment (when we know that all drivers needing to create a secure key derived from master HW key are done)

In the commit description you are explaining that this function is used to increment PRIBLOB but here is not not incremental this is a set to 0x3. To increment, you have to read, add one and write the PRIBLOB.

Emantor replied:
In https://github.com/OP-TEE/optee_os/pull/3608#discussion_r381135183 you said
> Hence, setting the PRIBLOB to 0x3 must remain a platform specific workaround done before switching to normal OS boot. It can not be part of CAAM driver at all.

Which is now implemented as a `plat-imx` specific workaround. I interpreted this as a request to remove this part from the CAAM driver completely. Please describe in detail what you had in mind for this change.

cneveux replied:
Problem with the proposed implementation is that it's add a new driver file in plat-imx directory and redefining stuff already done, like the base address of the register and the register offset. Work is duplicated and so if we have to fix/maintain we have to do it at different places.
Keeping function in the already existing caam driver (in the appropriate file) remove risk of bug and confusion during fix/maintenance/porting.
You can keep this operation as a workaround with a CFG_xxx flag in the CAAM driver itself making it in generic way (I mind transparently regardless device and if workaround must be applied).

The workaround concerns the CAAM driver hence it must be in CAAM driver.

HAL directory is here to maintain difference between i.MX and its help so much when porting driver between devices. Keeping the same logic even for workaround help also during porting to other devices. In other words, common driver logic/implementation doesn't have to be touched to add/fix specific devices, only hal has to be changed.

Emantor commented on core/arch/arm/plat-imx/crypto_conf.mk:
```diff
@@ -21,6 +21,11 @@ CFG_DBG_CAAM_TRACE ?= 0x2
 CFG_DBG_CAAM_BUF ?= 0x0
 
+# Enable the MKVB module used for the hardware unique key if the CAAM is enabled
+ifeq ($(CFG_NXP_CAAM),y)
```
will drop.

Emantor commented on core/drivers/crypto/caam/mkvb/caam_mkvb.c:
```diff
@@ -0,0 +1,83 @@
+
+static TEE_Result caam_get_mkvb(uint8_t *dest)
+{
+	struct caam_jobctx jobctx = { 0 };
```
will drop.

Emantor commented on core/drivers/crypto/caam/mkvb/caam_mkvb.c:
```diff
@@ -0,0 +1,83 @@
+
+	caam_desc_init(desc);
+	caam_desc_add_word(desc, DESC_HEADER(0));
+	caam_desc_add_word(desc, SEQ_OUT_PTR(32));
```
will fix.

Emantor commented:
I can't reply to some comments, quoting and replying here:
>  I still would like to have directory named blob and not mkvb. This is a caam blob operation.

Generic blob support would also be desirable for layerscape, but this will have conflicting definitions of `tee_otp_get_hw_unique_key`. Later commits can easily rename this (or even leave this and implement the other blob operations separately.

> instead of using MKVB_SIZE replace it with buf.length setup during the buffer buf allocation

done

> not needed

removed

> you need to ensure that hwkey->data size is not bigger than stored_key size

added a `COMPILE_TIME_ASSERT`

> Not sure that this commit is still necessary see that CFG_NXP_CAAM_RUNTIME_JR is not enabled if CFG_CRYPTO_DRIVER is not enabled and this is your case in the current PR

Only one of the changes is required, however I took the liberty of cleaning up the code to improve readability.

> I don't really like the crypto_driver_deinit function that could be prevent by adding a mvbk initialization function called by the ctrl_init function. The mvbk init will than generate the HUK and keep the value to be copied when tee_otp_get_hw_unique_key is called.
That way no need to add jrcfg as a global variable. Please review this implementation.

I have done that, will push a Review Commit shortly.

Emantor commented:
@jenswi-linaro please remove the stale label, I intend to update this come tuesday.

RNNord2 commented:
Hi. Using an approach like this PR, it seems possible to get the proper HUK from CAAM while avoiding the SoC hang due to Linux turning off the CAAM clocks. Is the intention with this approach also that the Fortuna PRNG should be enabled, and seeded by the CAAM TRNG during OP-TEE initialization/start-up? If so, are there any examples available to such a configuration and implementation? Furthermore, I wonder if this PR could be applied to OP-TEE OS 3.7.0, or if it has dependencies to a newer version?

Emantor commented:
> Hi. Using an approach like this PR, it seems possible to get the proper HUK from CAAM while avoiding the SoC hang due to Linux turning off the CAAM clocks. Is the intention with this approach also that the Fortuna PRNG should be enabled, and seeded by the CAAM TRNG during OP-TEE initialization/start-up? If so, are there any examples available to such a configuration and implementation? 

For now to avoid the clock issue, yes. The patches are rather simple, all that needs to be done is to save an RNG seed when the RNG is initialized and hand it over to OP-TEE once the `plat_rng_init()` is called. That used to be easier when the jobrings were disabled late, but storing the seed until the function is called and zeroing it explicitly afterwards should be fine.

> Furthermore, I wonder if this PR could be applied to OP-TEE OS 3.7.0, or if it has dependencies to a newer version?

Looking at the git log, the CAAM driver was ready in 3.7.0, so this PR might be backportable to an older version. I have no backport avilable unfortunately, but can give advice if things break.

RNNord2 commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -642,7 +643,7 @@ void caam_jr_resume(uint32_t pm_hint)
 				   jr_privdata->paddr_inrings,
 				   jr_privdata->paddr_outrings);
-#endif /* CFG_NXP_CAAM_RUNTIME_JR */
+	}
```
I may be missing the point, but doesn't this change means that the `#ifndef CFG_NXP_CAAM_RUNTUME_JR` a few lines up should be replaced with a `if (!IS_ENABLED(CFG_NXP_CAAM_RUNTIME_JR))` ?

Emantor replied:
This is a leftover commit which has AFAIR since been fixed by another commit, I will drop this on the next rebase.

Emantor commented:
> For commit "core: add early_driver initcalls"
> 
>     * Minor: typo in commit comment: s/radnomness/randomness/
I fixed that

>     * Nitpicking: indeed these new initcall levels can help, yet it may not be "driver" specific. Why not calling `early_init`? And really need a `_late` level related to this early init?

I renamed these to `early_init`. The caam driver wants a `late` variant which ensures that the jobrings have completed all ongoing jobs.

Emantor commented:
The update to increment the PRIBLOB now lives inside the CAAM HAL (under common, since these bits are AFAIK not SoC specific), but is only run if `CFG_CAAM_INC_PRIBLOB` is set. The function is  called directly after the MKVB initialization. I also fixed up some checkpatch issues and removed patches which are no longer necessary due to merged changes on master.

AFAICS the only change we do not disagree on is re-enabling `CFG_NXP_CAAM` and disabling `CFG_CRYPTO_DRIVER`. I still stand by my opinion that this is the correct way to go as we still don't have a OP-TEE compatible CAAM Linux driver and the HUK generation should not hinge upon its availability. As stated before I am fully in favor of removing the disabled `CFG_CRYPTO_DRIVER` as soon as it is compatible with the Linux kernel. But IMO providing a working HUK implementation for OP-TEE is an important enough change to enable by default in the next release.

Emantor commented:
cc/ @clementfaure & @cneveux

ldts commented:
@Emantor as you know the HUK provided by MKVB module is different for boards in Secure and Non-secure state: in non-secure state the HUK is not unique and seems to be common to each SoC family (with the marginal tests that I have run in 3 imx7ulp boards). 

It can happen that the RPMB key is written while the board is in non-secure state (we had an internal buggy release and we ended up burning two boards this way). Then on close, since the board is now secured, the HUK changes and RPMB is no longer accessible.

It is possible to work around this - just to avoid having to throw those boards to garbage - by using the HUK used while the board was in a non-secure state (lets call this the non-secure HUK).

So, the request is, could we always dump to the console the non-secure HUK provided by the MKVB? in an error scenario this can help recovering boards that required RPMB access to boot.

Emantor commented:
> @Emantor as you know the HUK provided by MKVB module is different for boards in Secure and Non-secure state: in non-secure state the HUK is not unique and seems to be common to each SoC family (with the marginal tests that I have run in 3 imx7ulp boards).

Indeed, the manual mentions that the test keys used in non-secure state are the same across the same class of devices.
> It can happen that the RPMB key is written while the board is in non-secure state (we had an internal buggy release and we ended up burning two boards this way - we forgot to enable SNVS after integrating the new CAAM driver). Then on close, since the board is now secured, the HUK changes and RPMB is no longer accessible.
> 
> It is possible to work around this - just to avoid having to throw those boards to garbage - by using the HUK used while the board was in a non-secure state (lets call this the non-secure HUK).

This looks to be difficult. The output of the MKVB depends on the used key and the CAAM derives the key state directly from the SNVS security state, so you can't force the CAAM to use the test keys in secure or trusted state AFAICS.

> So, the request is, could we always dump to the console the non-secure HUK provided by the MKVB? in an error scenario this can help recovering boards that required RPMB access to boot.

AFAICS it is not possible to switch the master key to the test key. The test key seems to only be used in non-secure state.

As a workaround for your devices, I'd suggest to disable the MKVB module added in this PR and instead providing your own `tee_otp_get_hw_unique_key()` which hardcodes the MKVB from a different device.
You can probably create a PR with a `CFG` for this, which hardcodes this for i.MX7ulp SoCs. I don't know whether upstream is interested in this however.

ldts commented:
@Emantor (let me help clarifying with an example)

See this workaround/hack to the issue I am mentioning (top two commits)
https://github.com/ldts/optee_os/commits/aeler-factory

the key so we can access RPMB would be passed as :
CFG_GET_ALTERNATIVE_HUK="0xc2, 0x0c, 0x77, 0xec, 0xad, 0x89, 0xdc, 0x96, 0xb7, 0x9f, 0xc8, 0xf7, 0xda, 0xab, 0x97, 0xb4,"

So we just need to know the default development key provided by the MKVB when the RPMB was first accessed when the board was still open..hence my suggestion of always putting in on the console while the board is not closed. That would prevent boards that require RPMB for early boot from being bricked (those in that state would just have to use a workaround along the linex of the one above)

Emantor commented:
Ah so what you suggest would be something like:
```
ifeq ($(CFG_RPMB_WRITE_KEY),y)
CFG_DBG_CAAM_BUF=0x400 # Bit(10) for MKVB
endif
```
Than the HuK is output when the RPMB key write is enabled as well.

ldts commented:
but wouldn't that expose the secret key as well? I think  the only situation where is useful for a developer  and safe for the product to know the HUK used to generate the RPMB key is if the HUK was the infact the **test key....never the one used in the secure state**. So if someone by mistake disables the SNVS (so the plat_rpmb_is_ready check passes when it should not), and the test key is used to derive the RPMB key and RPMB is initialized.... and then the board is put in secure state, we will know how to recover the RPMB access using the previously dumped test key.

ldts commented:
so I perhaps would do something like the following

1) make sure rpmb_is_ready is properly implemented (or it can lead to bricked boards)

```
bool __weak plat_rpmb_is_ ready()
{  return false }
```

2) then in imx-huk perhaps have the following in case we ever need to use this key

```
tee_otp_get_hw_unique_key () 
                 if (!plat_rpmb_is_ready())
                        DHEXDUMP(huk...)
```

Emantor commented:
In the current patchset, the HUK generation is fully contained within the MKVB module, there is not `imx-huk` file any longer. I don't really see the benefit of adding the rpmb check inside the CAAM driver. With `CFG_DBG_CAAM_BUF = 0x400` you'll get the key printed either way and can configure this for your platform. Since the RPMB bringup needs two OP-TEE OS images either way, since you don't want images which have `CFG_RPMB_WRITE_KEY=y` in production outside the bringup environment, I think this can be solved using configuration locally.

ldts commented:
true, defaulting that DBG config is enough. awesome

ldts commented:
yes.
so why not just have this upstream?

ifneq ($(CFG_RPMB_WRITE_KEY), y)
CFG_DBG_CAAM_BUF ?= 0x400 # Bit(10) for MKVB
endif

Emantor commented:
``
ifneq ($(CFG_RPMB_WRITE_KEY), y)
CFG_DBG_CAAM_BUF ?= 0x400 # Bit(10) for MKVB
endif
```
Correct me if I am wrong, but this will always print the HUK even if `CFG_RPMB_WRITE_KEY` is not set, right? I don't think its advisable, even though `CFG_TEE_CORE_DEBUG` needs AFAIR to be set as well.

etienne-lms commented on core/drivers/crypto/caam/include/caam_trace.h:
```diff
@@ -194,4 +195,27 @@
+#define MKVB_DUMPDESC(desc)                                                    \
+	do {                                                                   \
+		MKVB_TRACE("MKVB Descriptor");                                 \
+		DRV_DUMPDESC(desc);                                            \
```
style: prefer replace spaces with tabulations

etienne-lms commented on core/drivers/crypto/caam/mkvb/caam_mkvb.c:
```diff
@@ -0,0 +1,79 @@
+{
+	struct caam_jobctx jobctx = { };
+	enum caam_status res = CAAM_NO_ERROR;
+	struct caambuf buf = {};
```
s/`{}`/`{ }`/

Emantor commented:
Changes:
- applied Etiennes Tags
- renamed MKVB to BLOB
- removed the configuration change

This should allow this PR to be unblocked and make it ready for merge. cc @cneveux & @clementfaure

etienne-lms commented on core/drivers/crypto/caam/blob/caam_blob.c:
```diff
@@ -0,0 +1,79 @@
+#include <caam_common.h>
+#include <caam_hal_ctrl.h>
+#include <caam_jr.h>
+#include <caam_blob.h>
```
minor: fix ordering (move this one to the 1st place)

etienne-lms commented on core/drivers/crypto/caam/blob/caam_blob.c:
```diff
@@ -0,0 +1,79 @@
+#include <string.h>
+#include <tee/cache.h>
+
+#define	MKVB_SIZE	32
```
prefer no tab between `#define` and macro name.

etienne-lms commented on core/drivers/crypto/caam/blob/caam_blob.c:
```diff
@@ -0,0 +1,79 @@
+{
+	struct caam_jobctx jobctx = { };
+	enum caam_status res = CAAM_NO_ERROR;
+	struct caambuf buf = {};
```
s/`{}`/`{ }`/

etienne-lms commented on core/drivers/crypto/caam/blob/caam_blob.c:
```diff
@@ -0,0 +1,79 @@
+
+	if (res != CAAM_NO_ERROR) {
+		BLOB_TRACE("JR return code: %"PRIx32, res);
+		BLOB_TRACE("MKVB failed: Job status 0x%08" PRIx32,
```
for both: suggest `"%#"PRIx32` to get a nice `0x` prefixed 32bit hexadecimal trace.

etienne-lms replied:
FYI, no need space between `%` and `PRI...`.
Same if there are other flag characters between `%` and `PRI...`, i.e. `"%"PRIu8`, `"%#08"PRIx32`.
I think it is better without a space but that's not mandated.

clementfaure replied:
Alright!

etienne-lms commented on core/drivers/crypto/caam/blob/caam_blob.c:
```diff
@@ -0,0 +1,79 @@
+	enum caam_status res = CAAM_NO_ERROR;
+	struct caambuf buf = {};
+	uint32_t *desc = NULL;
+
```
suggest `assert(!mkvb_retrieved);`

etienne-lms commented on core/drivers/crypto/caam/blob/caam_blob.c:
```diff
@@ -0,0 +1,79 @@
+
+	desc = caam_calloc_desc(8);
+	if (!desc)
+		goto out_buf;
```
missing `res = CAAM_NO_MEMORY` or like.

etienne-lms commented on core/drivers/crypto/caam/include/caam_blob.h:
```diff
@@ -0,0 +1,24 @@
+ */
+enum caam_status caam_blob_mkvb_init(vaddr_t baseaddr);
+#else
+static inline enum caam_status caam_blob_mkvb_init(vaddr_t baseaddr)
```
add `__unused` after `baseaddr`.

You should try a build with `CFG_NXP_CAAM_BLOB_DRV=n` to test.

etienne-lms commented:
Previous comment series apply to commit "caam: add BLOB module for MKVB retrieval".

clementfaure commented on core/drivers/crypto/caam/include/caam_blob.h:
```diff
@@ -0,0 +1,24 @@
+{
+	return CAAM_NO_ERROR;
+}
+#endif /* CFG_NXP_CAAM_MKVB_DRV */
```
```CFG_NXP_CAAM_BLOB_DRV``` instead of ```CFG_NXP_CAAM_MKVB_DRV ```

clementfaure commented on core/drivers/crypto/caam/hal/common/hal_ctrl.c:
```diff
@@ -57,3 +60,38 @@ uint8_t caam_hal_ctrl_era(vaddr_t baseaddr)
+	else if (val == 1)
+		blob = val + 2;
+	else
+		panic("Error locking PRIBLOB, PRIBLOB >=3");
```
```PRIBLOB =3 ```

Emantor commented:
Thanks for the reviews @etienne-lms & @clementfaure, I have pushed review commits for your comments.

clementfaure commented on core/drivers/crypto/caam/include/caam_trace.h:
```diff
@@ -35,6 +35,7 @@
 #define DBG_TRACE_HASH	 BIT32(7)  /* Hash trace */
 #define DBG_TRACE_RSA	 BIT32(8)  /* RSA trace */
 #define DBG_TRACE_CIPHER BIT32(9)  /* Cipher dump Buffer */
+#define DBG_TRACE_BLOB   BIT32(10) /* BLOB trace */
```
Could you also report this value in comment in ```crypto_conf.mk``` ?
https://github.com/OP-TEE/optee_os/blob/9de9e81542ca568699c81220b563d257d430899b/core/arch/arm/plat-imx/crypto_conf.mk#L11-L23

Emantor replied:
Done in the latest review commit.

Emantor commented:
Rebased, squashed, tags applied.

jforissier commented:
@Emantor there are a couple of trivial conflicts with the CAAM HMAC patches merged minutes ago, could you please rebase? Thanks.

Emantor commented:
Done.
